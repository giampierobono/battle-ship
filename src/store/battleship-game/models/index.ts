export * from './set-game-player-turn-payload.model';
export * from './save-player-move-payload.model';
export * from './set-winner-player-payload.model';
