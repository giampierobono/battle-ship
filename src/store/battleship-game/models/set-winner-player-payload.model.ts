export interface SetWinnerPlayerPayloadModel {
  winnerPlayerIndex: number;
}
