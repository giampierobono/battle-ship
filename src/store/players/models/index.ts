export * from './add-players-payload.model';
export * from './create-player-payload.model';
