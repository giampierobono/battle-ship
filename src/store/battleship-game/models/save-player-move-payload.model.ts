import { PositionStateEnum, ShotPositionModel } from '../../../models';

export interface SavePlayerMovePayloadModel {
  shotPosition: ShotPositionModel;
  state: PositionStateEnum;
}
