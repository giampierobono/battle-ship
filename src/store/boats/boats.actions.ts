import { createAction, props } from '@ngrx/store';
import { AddBoatsToPlayerPayload, HitPlayerBoatPayloadModel } from './models';

export const addBoatsToPlayer = createAction('[Boats] Add boat to player', props<AddBoatsToPlayerPayload>());

export const hitPlayerBoat = createAction('[Boats] Hit player boat', props<HitPlayerBoatPayloadModel>());
