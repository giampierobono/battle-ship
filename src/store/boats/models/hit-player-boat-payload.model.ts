import { ShotPositionModel } from '../../../models';

export interface HitPlayerBoatPayloadModel {
  boatIndex: number;
  position: ShotPositionModel;
}
