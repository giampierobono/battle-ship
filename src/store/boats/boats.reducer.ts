import { createReducer } from '@ngrx/store';
import * as BoatsActions from './boats.actions';
import { BoatModel, PositionStateEnum } from '../../models';
import { produceOn } from '../common/utils';

export interface IBoatsState {
  boats: Array<Array<BoatModel>>;
}

export const boatsInitialState: IBoatsState = {
  boats: [],
};

export const boatsReducer = createReducer(
  boatsInitialState,
  produceOn(BoatsActions.addBoatsToPlayer, (state, { boats }) => {
    state.boats = boats;
  }),
  produceOn(BoatsActions.hitPlayerBoat, (state, { position: { playerIndex, row, column }, boatIndex }) => {
    state.boats[playerIndex][boatIndex].positions[row][column] = PositionStateEnum.HIT;
    state.boats[playerIndex][boatIndex].hits++;
  })
);
