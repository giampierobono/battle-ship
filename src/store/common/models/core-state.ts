import { IPlayersState } from '../../players';
import { IBoatsState } from '../../boats';
import { IBoatShapeState } from '../../boats-shapes';
import { IGameSettingsState } from '../../game-settings';
import { IBattleShipGameState } from '../../battleship-game';

export interface CoreState {
  players: IPlayersState;
  boats: IBoatsState;
  boatShapes: IBoatShapeState;
  settings: IGameSettingsState;
  game: IBattleShipGameState;
}
