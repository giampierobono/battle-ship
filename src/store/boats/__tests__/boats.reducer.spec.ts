import { boatsInitialState, boatsReducer } from '../boats.reducer';
import { addBoatsToPlayer, hitPlayerBoat } from '../boats.actions';
import { BoatModel, PositionStateEnum } from '../../../models';

describe('Boats reducer', () => {
  it('should return initial state for unknown action', () => {
    expect(boatsReducer(boatsInitialState, { type: 'UNKNOWN' })).toEqual(boatsInitialState);
  });

  it('should add boat to player', () => {
    const newState = boatsReducer(
      boatsInitialState,
      addBoatsToPlayer({ boats: [[{ playerIndex: 0, boatIndex: 0, shapeKey: 0, hits: 0, positions: {} }]] })
    );
    expect(newState.boats.length).toBe(1);
    expect(newState.boats).toMatchSnapshot();
  });

  it('should mark boat position as HIT and increase hits count', () => {
    const stateBoats: Array<Array<BoatModel>> = [
      [{ positions: { 0: { 0: PositionStateEnum.BOAT } }, boatIndex: 0, playerIndex: 0, hits: 1, shapeKey: 2 }],
    ];
    const newState = boatsReducer(
      { ...boatsInitialState, boats: stateBoats },
      hitPlayerBoat({ boatIndex: 0, position: { playerIndex: 0, column: 0, row: 0 } })
    );
    expect(newState.boats[0][0].positions[0][0]).toBeTruthy();
    expect(newState.boats[0][0].hits).toBe(2);
  });
});
