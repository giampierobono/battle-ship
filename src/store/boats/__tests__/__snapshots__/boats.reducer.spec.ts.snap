// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boats reducer should add boat to player 1`] = `
Array [
  Array [
    Object {
      "boatIndex": 0,
      "hits": 0,
      "playerIndex": 0,
      "positions": Object {},
      "shapeKey": 0,
    },
  ],
]
`;
